### PR TITLE
8328 nnunet bundle integration

### LINKED ISSUE #8328: nnUNet Integration for MONAI Bundle
To better align nnUNet with the MONAI ecosystem, including tools like MONAI Label and MONAIAlgo for FL in NVFlare, it would be highly beneficial to support wrapping nnUNet models within a MONAI Bundle. This integration would better integrate nnUNet models with other MONAI ecosystem applications—like MONAI Label and MONAIAlgo for FL in NVFlare—which already support the MONAI Bundle. With this enhancement, any model trained natively using nnUNet could be converted into a MONAI Bundle for seamless deployment at inference time in downstream tasks.

Additionally, this bundle could also be used during training, providing an alternative to the current native nnUNet and nnUNetV2Runner implementations in MONAI.

### PR DESCRIPTION
**Pull request by SimoneBendazzoli93**
Fixes #8328 

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] Breaking change (fix or new feature that would cause existing functionality to change).
- [X] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by SimoneBendazzoli93** [comment]
Hi @ericspod , I have now moved the nnUNet bundle code under `monai.apps.nnunet`

File: monai/apps/nnunet/nnunet_bundle.py [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+        if fold != "all":
+            try:
+                fold = int(fold)
+            except ValueError as e:
+                print(
+                    f'Unable to convert given value for fold to int: {fold}. fold must bei either "all" or an integer!'
+                )
+                raise e
+
+    if int(num_gpus) > 1:
```
Comment (KumoLiu) [inline monai/apps/nnunet/nnunet_bundle.py:120]: It seems we don't need support `num_gpus>1` in this `get_nnunet_trainer` function?

File: monai/apps/nnunet/nnunet_bundle.py [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+    This class integrates nnUNet model with MONAI framework by loading necessary configurations,
+    restoring network architecture, and setting up the predictor for inference.
+    """
+
+    def __init__(self, predictor, model_folder, model_name="model.pt"):
+        super().__init__()
+        self.predictor = predictor
+
+        model_training_output_dir = model_folder
+        use_folds = ["0"]
```
Comment (KumoLiu) [inline monai/apps/nnunet/nnunet_bundle.py:184]: why hard code it here? what's the meaning of this `use_folds`?

File: monai/apps/nnunet/nnunet_bundle.py [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+        predictor.list_of_parameters = parameters
+        predictor.network = network
+        predictor.dataset_json = dataset_json
+        predictor.trainer_name = trainer_name
+        predictor.allowed_mirroring_axes = inference_allowed_mirroring_axes
+        predictor.label_manager = plans_manager.get_label_manager(dataset_json)
+        if (
+            ("nnUNet_compile" in os.environ.keys())
+            and (os.environ["nnUNet_compile"].lower() in ("true", "1", "t"))
+            # and not isinstance(predictor.network, OptimizedModule)
```
Comment (KumoLiu) [inline monai/apps/nnunet/nnunet_bundle.py:252]: may i ask the reason for this comment here?

File: monai/apps/nnunet/nnunet_bundle.py [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+        ):
+            print("Using torch.compile")
+            # predictor.network = torch.compile(self.network)
```
Comment (KumoLiu) [inline monai/apps/nnunet/nnunet_bundle.py:255]: It seems this line need to be uncommented otherwise will not using `torch.compile`

File: monai/apps/nnunet/nnunet_bundle.py [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+        #    image_or_list_of_images = []
+
+        # for img in x:
+        # if isinstance(img, MetaTensor):
+        #    properties_or_list_of_properties.append({"spacing": img.meta['pixdim'][0][1:4].numpy().tolist()})
+        #    image_or_list_of_images.append(img.cpu().numpy()[0,:])
+        # else:
+        #    raise TypeError("Input must be a MetaTensor or a tuple of MetaTensors.")
+
+        # else:  # if batch is collated
```
Comment (KumoLiu) [inline monai/apps/nnunet/nnunet_bundle.py:293]: If this comment code are for test, may need remove them, thanks.

File: monai/apps/nnunet/nnunet_bundle.py [hunk 6]
```diff
@@ -0,0 +1,448 @@
...
+        # prediction_output is a list of numpy arrays, with dimensions (H, W, D), output from ArgMax
+
+        out_tensors = []
+        for out in prediction_output:  # Add batch and channel dimensions
+            out_tensors.append(torch.from_numpy(np.expand_dims(np.expand_dims(out, 0), 0)))
+        out_tensor = torch.cat(out_tensors, 0)  # Concatenate along batch dimension
+
+        # if type(x) is tuple:
+        #    return MetaTensor(out_tensor, meta=x[0].meta)
+        # else:
```
Comment (KumoLiu) [inline monai/apps/nnunet/nnunet_bundle.py:323]: Also here.

**Review by KumoLiu — COMMENTED** [review]
Thanks for the contribution!
It seems there still several places need to be polished. Leave several comments in line.

File: monai/apps/nnunet/nnunet_bundle.py [hunk 7]
```diff
@@ -0,0 +1,448 @@
...
+        Whether to continue training from a checkpoint. Default is False.
+    only_run_validation : bool, optional
+        Whether to only run validation. Default is False.
+    disable_checkpointing : bool, optional
+        Whether to disable checkpointing. Default is False.
+    val_with_best : bool, optional
+        Whether to validate with the best model. Default is False.
+    device : str, optional
+        The device to be used for training. Default is 'cuda'.
+    pretrained_model : str, optional
```
[the remark this answers is not in the document]
Comment (SimoneBendazzoli93) [reply monai/apps/nnunet/nnunet_bundle.py]: I have now removed `pretrained_weights` and left only `pretrained_model`

File: monai/apps/nnunet/nnunet_bundle.py (hunk 2 above)
Comment (SimoneBendazzoli93) [reply monai/apps/nnunet/nnunet_bundle.py]: Same as above

File: monai/apps/nnunet/nnunet_bundle.py (hunk 3 above)
Comment (SimoneBendazzoli93) [reply monai/apps/nnunet/nnunet_bundle.py]: These comments are also from the original nnUNet implemenation. Now they are removed

File: monai/apps/nnunet/nnunet_bundle.py [hunk 8]
```diff
@@ -0,0 +1,448 @@
...
+from monai.utils import optional_import
+
+join, _ = optional_import("batchgenerators.utilities.file_and_folder_operations", name="join")
+load_json, _ = optional_import("batchgenerators.utilities.file_and_folder_operations", name="load_json")
+
+__all__ = ["get_nnunet_trainer", "get_nnunet_monai_predictor", "convert_nnunet_to_monai_bundle", "ModelnnUNetWrapper"]
+
+
+def get_nnunet_trainer(
+    dataset_name_or_id,
```
[the remark this answers is not in the document]
Comment (SimoneBendazzoli93) [reply monai/apps/nnunet/nnunet_bundle.py]: type hint added for all the functions in this file

File: monai/apps/nnunet/nnunet_bundle.py [hunk 9]
```diff
@@ -0,0 +1,448 @@
...
+    ----------
+    dataset_name_or_id : Union[str, int]
+        The name or ID of the dataset to be used.
+    configuration : str
+        The configuration name for the training.
+    fold : Union[int, str]
+        The fold number or 'all' for cross-validation.
+    trainer_class_name : str, optional
+        The class name of the trainer to be used. Default is 'nnUNetTrainer'.
+    plans_identifier : str, optional
```
[the remark this answers is not in the document]
Comment (SimoneBendazzoli93) [reply monai/apps/nnunet/nnunet_bundle.py:86]: For a list of supported options, reference to https://github.com/MIC-DKFZ/nnUNet/tree/master/nnunetv2/training/nnUNetTrainer/variants

File: monai/apps/nnunet/nnunet_bundle.py [hunk 10]
```diff
@@ -238,30 +227,23 @@ def __init__(self, predictor, model_folder, model_name="model.pt"):
...
+        predictor.configuration_manager = configuration_manager  # type: ignore
+        predictor.list_of_parameters = parameters  # type: ignore
+        predictor.network = network  # type: ignore
+        predictor.dataset_json = dataset_json  # type: ignore
+        predictor.trainer_name = trainer_name  # type: ignore
+        predictor.allowed_mirroring_axes = inference_allowed_mirroring_axes  # type: ignore
+        predictor.label_manager = plans_manager.get_label_manager(dataset_json)  # type: ignore
+        if ("nnUNet_compile" in os.environ.keys()) and (os.environ["nnUNet_compile"].lower() in ("true", "1", "t")):
             print("Using torch.compile")
-            # predictor.network = torch.compile(self.network)
```
[the remark this answers is not in the document]
Comment (SimoneBendazzoli93) [reply monai/apps/nnunet/nnunet_bundle.py]: This is also from the original nnUNet implementation: https://github.com/MIC-DKFZ/nnUNet/blob/ea8baa328fba93dd41f8001675557174c8923c41/nnunetv2/training/nnUNetTrainer/nnUNetTrainer.py#L218

To enable it, you need to specify the `nnUNet_compile` var, and then the function called is `self.network = torch.compile(self.network)`

For now, I decided to disable this option, but it can definitely be included.

**Comment by KumoLiu** [comment]
Hi @ericspod, let me know if you have any other concerns on this PR, thanks!

**Review by ericspod — APPROVED** [review]
Looks good to me now!
